neuralforecast/models/vcformer.py

```diff
@@ -0,0 +1,477 @@
+        self.conv2 = nn.Conv1d(in_channels=d_ff, out_channels=d_model, kernel_size=1)
+        self.norm1 = nn.LayerNorm(d_model)
+        self.norm2 = nn.LayerNorm(d_model)
+        self.norm3 = nn.LayerNorm(d_model)
+        self.dropout = nn.Dropout(dropout)
+        self.activation = F.relu if activation == "relu" else F.gelu
+
+    def forward(self, x, attn_mask=None, tau=None, delta=None):
+        # VCA module
+        new_x, attn = self.attention(x, x, x, attn_mask=attn_mask, tau=tau, delta=delta)
+        x = x + self.dropout(new_x)
+        y = x = self.norm1(x)
+
+        # KTD Module
+        x_ktd = y
+        y = self.dropout(self.ktd(x_ktd))
+
+        return self.norm2(x_ktd + y), attn
+
+# %% ../../nbs/models.vcformer.ipynb 14
+class TriangularCausalMask:
+    def __init__(self, B, L, device="cpu"):
+        mask_shape = [B, 1, L, L]
+        with torch.no_grad():
+            self._mask = torch.triu(
+                torch.ones(mask_shape, dtype=torch.bool), diagonal=1
+            ).to(device)
+
+    @property
+    def mask(self):
+        return self._mask
+
+
+class VarCorAttention(nn.Module):
+    def __init__(
+        self,
+        mask_flag=True,
+        factor=5,
+        scale=None,
+        attention_dropout=0.1,
+        output_attention=False,
+    ) -> None:
+        super(VarCorAttention, self).__init__()
+
+        self.scale = scale
+        self.mask_flag = mask_flag
+        self.output_attention = output_attention
+        self.dropout = nn.Dropout(attention_dropout)
+
+    def origin_compute_cross_cor(self, queries, keys):
+        q_fft = torch.fft.rfft(queries, dim=-1)
+        k_fft = torch.fft.rfft(keys, dim=-1)
+
+        res = q_fft * k_fft
+        corr = torch.fft.irfft(res, dim=-1)
+        corr = corr.mean(dim=-1)
+        return corr
+
+    def compute_cross_cor(self, queries, keys):
+
+        q_fft = torch.fft.rfft(queries, dim=-1)
+        k_fft = torch.fft.rfft(keys, dim=-1)
+        q_fft = q_fft.unsqueeze(1)  # [D,1,T/2+1]
+        k_fft = torch.conj(k_fft.unsqueeze(0))  # [1,D,T/2+1]
+        res = q_fft * k_fft  # [D,D,T/2+1]
+        corr = torch.fft.irfft(res, dim=-1)
+        corr = corr.mean(dim=-1)
+
+        return corr
+
+    def optimized_compute_cross_cor(self, queries, keys):
+        # Perform batched FFT
+        q_fft = torch.fft.rfft(queries, dim=-1)
+        k_fft = torch.fft.rfft(keys, dim=-1)
+
+        # Expand dimensions for broadcasting
+        q_fft = q_fft.unsqueeze(2)  # [B, D, 1, T/2+1]
+        k_fft = torch.conj(k_fft.unsqueeze(1))  # [B, 1, D, T/2+1]
+
+        # Element-wise multiplication and batched inverse FFT
+        res = q_fft * k_fft  # [B, D, D, T/2+1]
+        corr = torch.fft.irfft(res, dim=-1)
+
+        # Mean across the time dimension
+        corr = corr.mean(dim=-1)
+
+        return corr
+
+    def forward(self, queries, keys, values, attn_mask, tau=None, delta=None):
```

Comment:
This seems quite inefficient, don't know if this is the original code but I think we could refactor this ourselves, i.e. get rid of the for-loops and replace the fft cross-correlation operator by a normal conv1d.